packages/yarnpkg-pnpify/sources/buildNodeModulesTree.ts

```diff
@@ -0,0 +1,274 @@
+};
+
+
+/**
+ * Converts hoisted tree to node modules map
+ *
+ * @param pnp PnP API
+ * @param hoistedTree hoisted package tree from `RawHoister`
+ * @param locators locators
+ *
+ * @returns node modules map
+ */
+const populateNodeModulesTree = (pnp: PnpWalkApi, hoistedTree: HoistedTree, locators: PackageLocator[], options: NodeModulesTreeOptions): NodeModulesTree => {
+  const tree: NodeModulesTree = new Map();
+
+  // Hard link Regex heuristics, needed for old PnP API versions support
+  const HARD_LINK_REGEX = /((\.zip|-integrity)[\\/]node_modules[\\/]|-virtual-[0-9a-f]{10})/;
+
+  const NO_DEPS = new Set<HoisterPackageId>();
+  const getLocation = (locator: PackageLocator): [PortablePath, LinkType] => {
+    const info = pnp.getPackageInformation(locator)!;
+    if (pnp.resolveVirtual) {
+      const truePath = (!options.pnpifyFs && locator.reference && locator.reference.startsWith('virtual:')) ? pnp.resolveVirtual(info.packageLocation) : info.packageLocation;
+      return [npath.toPortablePath(truePath || info.packageLocation), info.linkType];
+    } else {
+      const linkType = info.linkType || HARD_LINK_REGEX.test(info.packageLocation) ? LinkType.HARD : LinkType.SOFT;
+      if (linkType === LinkType.SOFT && locator.reference && locator.reference!.indexOf('virtual:') === 0) {
+        const realLocator = {name: locator.name!, reference: locator.reference!.split('#')[1]};
+        const realInfo = pnp.getPackageInformation(realLocator)!;
+        if (!realInfo) {
+          return [npath.toPortablePath(info.packageLocation), LinkType.HARD];
+        } else {
+          return [npath.toPortablePath(realInfo.packageLocation), linkType];
+        }
+      } else {
+        return [npath.toPortablePath(info.packageLocation), linkType];
+      }
+    }
+  };
+  const getPackageName = (locator: PackageLocator): { name: Filename, scope: Filename | null } => {
+    const [nameOrScope, name] = locator.name!.split('/');
+    return name ? {scope: toFilename(nameOrScope), name: toFilename(name)} : {scope: null, name: toFilename(nameOrScope)};
+  };
+
+  const seenPkgIds = new Set();
+  const buildTree = (nodeId: HoisterPackageId, prefix: PortablePath) => {
+    seenPkgIds.add(nodeId);
+    const depIds = hoistedTree[nodeId] || NO_DEPS;
+    for (const depId of depIds) {
+      const locator = locators[depId];
+      const {name, scope} = getPackageName(locator);
+      const packageNameParts = scope ? [scope, name] : [name];
+      const nodeModulesDirPath = ppath.join(prefix, NODE_MODULES);
+      let depPrefix = ppath.join(...([nodeModulesDirPath].concat(packageNameParts)));
+      if (!seenPkgIds.has(depId)) {
+        const nodeVal = getLocation(locator);
+        tree.set(depPrefix, nodeVal);
+
+        const segments = depPrefix.split(NODE_MODULES_SUFFIX);
+        let segCount = segments.length - 1;
+        while (segCount > 0) {
+          const nodePath = npath.toPortablePath(segments.slice(0, segCount).join(NODE_MODULES_SUFFIX));
```

Comment:
As we discussed on Discord I think a single loop on the path split on `/` could work and make the code simpler

Reply:
Done